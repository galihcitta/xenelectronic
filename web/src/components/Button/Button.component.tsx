import React from 'react';

import { ButtonWrapper } from './Button.styles';

export interface ButtonProps {
  disabled?: boolean;
  onClick?: () => void;
  btnType?: string;
  type?: 'button' | 'submit' | 'reset' | undefined;
  children: any;
}

export const Button: React.FC<ButtonProps> = ({
  disabled,
  onClick,
  btnType,
  type = 'button',
  children
}) => (
  <ButtonWrapper
    disabled={disabled}
    onClick={onClick}
    className={[btnType].join(' ')}
    type={type}
  >
    {children}
  </ButtonWrapper>
);
