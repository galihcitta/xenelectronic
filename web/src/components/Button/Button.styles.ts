import styled from 'styled-components';

import { color, fontWeight } from '../../utils/styles';

export const ButtonWrapper = styled.button`
  width: 100%;
  border: 1.5px solid ${color.black};
  background: ${color.black};
  color: ${color.white};
  padding: 0.8em 1.9em;
  text-transform: uppercase;
  font-weight: ${fontWeight.semiBold};
  font-size: 0.9rem;
  cursor: pointer;
  transition: all 0.15s ease-in-out;

  &:hover {
    background: ${color.white};
    color: ${color.black};
  }

  &:disabled {
    border: 1.5px solid ${color.gray};
    background: ${color.gray};
    opacity: 0.55;
    cursor: not-allowed;

    &:hover {
      background: ${color.gray};
      color: ${color.white};
    }
  }
`;
