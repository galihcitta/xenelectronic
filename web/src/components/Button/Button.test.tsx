import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import { Button, ButtonProps } from './Button.component';

describe('<Button />', () => {
  const mockOnClick = jest.fn();

  const props: ButtonProps = {
    onClick: mockOnClick,
    children: 'Submit',
    type: 'button'
  };

  const setup = (props: ButtonProps) => {
    const utils = render(<Button {...props} />);
    const button = utils.container.querySelector('button') as HTMLButtonElement;
    return { ...utils, button };
  };

  it('renders correctly', () => {
    const { asFragment } = setup(props);
    expect(asFragment()).toMatchSnapshot();
  });

  it('contains disabled attribute', () => {
    const { button } = setup({ ...props, disabled: true });
    expect(button).toHaveAttribute('disabled', '');
  });

  it('calls `onClick` function', () => {
    const { button } = setup(props);
    fireEvent.click(button);
    expect(mockOnClick).toBeCalledTimes(1);
  });

  it('contains `dark` class', () => {
    const { button } = setup({ ...props, btnType: 'dark' });
    expect(button).toHaveClass('dark');
  });

  it('contains `submit` type', () => {
    const { button } = setup({ ...props, type: 'submit' });
    expect(button).toHaveAttribute('type', 'submit');
  });
});
