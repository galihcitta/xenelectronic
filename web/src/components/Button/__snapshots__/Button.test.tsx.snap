// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Button /> renders correctly 1`] = `
<DocumentFragment>
  <button
    class="sc-bdVaJa cPICdt"
    type="button"
  >
    Submit
  </button>
</DocumentFragment>
`;
